src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/FileBasedPrograms/CanonicalMiscFilesProjectLoader.cs

```diff
@@ -138,7 +136,28 @@ await _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory.ApplyChangeToW
         }, cancellationToken);
     }
 
-    private async ValueTask<TextDocument> AddDocumentToExistingProject_NoLockAsync(string documentPath, SourceText documentText, ProjectId existingProjectId, CancellationToken cancellationToken)
+    private async ValueTask<TextDocument> ForkCanonicalProjectAndAddDocument_NoLockAsync(string documentPath, SourceText documentText, CancellationToken cancellationToken)
+    {
+        var newProjectId = ProjectId.CreateNewId(debugName: $"Forked Misc Project for '{documentPath}'");
+        var newDocumentInfo = DocumentInfo.Create(
+            DocumentId.CreateNewId(newProjectId),
+            name: Path.GetFileName(documentPath),
+            loader: TextLoader.From(TextAndVersion.Create(documentText, VersionStamp.Create())),
+            filePath: documentPath);
+
+        var forkedProjectInfo = await GetForkedProjectInfo_NoLockAsync(newDocumentInfo, documentText, cancellationToken);
+
+        await _workspaceFactory.MiscellaneousFilesWorkspaceProjectFactory.ApplyChangeToWorkspaceAsync(workspace =>
+        {
+            workspace.OnProjectAdded(forkedProjectInfo);
+        }, cancellationToken);
```

Comment:
What are the semantics of cancellation here?  I definitely don't know what the intent is, or if it is ok that workspace mutation ops may not happen. @dibarbet what the general intuition here?

Reply:
This cancellation token here would be cancelled if the request gets cancelled by the client.  Assuming that cancelling the `ApplyChangeToWorkspaceAsync` doesn't leave the workspace in an intermediate state, a cancellation here is fine.  We won't add the project to the workspace, and that is totally fine.  If another call comes in later asking about this document, we'll get here again and try and add it to the workspace.

That applies to `AddDocumentToPrimordialProject_NoLockAsync` as well.  For `CreatePrimordialProjectAndAddDocument_NoLock`, that is not passed a token as it initiates a project loading operation behind the scenes which should not be cancelled even if the request itself is cancelled, as we'll re-use it for other requests.